File: examples/javascript/selenium/tests/visual.test.js
const { Builder, By, until } = require('selenium-webdriver');
const { witness } = require('../testivai-witness');

describe('Visual Regression Tests', () => {
  let driver;

  beforeAll(async () => {
    // Configure Chrome with remote debugging port for TestivAI CDP
    const chromeOptions = {
      'goog:chromeOptions': {
        args: ['--remote-debugging-port=9222', '--headless']
      }
    };
    
    driver = await new Builder()
      .forBrowser('chrome')
      .withCapabilities(chromeOptions)
      .build();
  });

  afterAll(async () => {
    if (driver) {
      await driver.quit();
    }
  });

  test('homepage loads and looks correct', async () => {
    await driver.get('http://localhost:3000');
    await driver.wait(until.titleContains('TestivAI'), 5000);
    await witness(driver, 'homepage');
  });

  test('navigation bar is consistent', async () => {
    await driver.get('http://localhost:3000');
    await driver.wait(until.elementLocated(By.css('nav')), 5000);
    await witness(driver, 'navbar');
  });

  test('product cards render correctly', async () => {
    await driver.get('http://localhost:3000');
    // Scroll to product cards section
    const productSection = await driver.wait(
      until.elementLocated(By.css('.product-cards, [data-testid="product-cards"]')),
      5000
    );
    await driver.executeScript('arguments[0].scrollIntoView(true);', productSection);
    await driver.sleep(500); // Wait for any lazy loading
    await witness(driver, 'product-cards');
  });

  test('checkout form layout', async () => {
    await driver.get('http://localhost:3000/checkout');
    await driver.wait(until.elementLocated(By.css('form')), 5000);
    await witness(driver, 'checkout-form');
  });

  test('checkout form with validation', async () => {
    await driver.get('http://localhost:3000/checkout');
    
    // Fill partial form to trigger validation
    const emailInput = await driver.wait(until.elementLocated(By.css('#email, [name="email"]')), 5000);
    await emailInput.sendKeys('test@');
    
    const submitButton = await driver.findElement(By.css('button[type="submit"], [data-testid="submit"]'));
    await submitButton.click();
    
    // Wait for validation messages
    await driver.sleep(1000);
    await witness(driver, 'checkout-validation');
  });
});

File: examples/java/selenium-junit5/src/test/java/ai/testivai/tests/VisualTest.java
package ai.testivai.tests;

import static org.assertj.core.api.Assertions.assertThat;

import java.time.Duration;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import ai.testivai.helpers.TestivAIWitness;
import io.github.bonigarcia.wdm.WebDriverManager;

public class VisualTest {
    
    private static WebDriver driver;
    private static WebDriverWait wait;
    
    @BeforeAll
    static void setup() {
        // Setup Chrome driver with remote debugging for TestivAI
        WebDriverManager.chromedriver().setup();
        
        ChromeOptions options = new ChromeOptions();
        options.addArguments("--remote-debugging-port=9222");
        options.addArguments("--headless");
        
        driver = new ChromeDriver(options);
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
    }
    
    @AfterAll
    static void teardown() {
        if (driver != null) {
            driver.quit();
        }
    }
    
    @Test
    void homepageLooksCorrect() {
        driver.get("http://localhost:3000");
        wait.until(ExpectedConditions.titleContains("TestivAI"));
        
        TestivAIWitness.witness(driver, "homepage");
    }
    
    @Test
    void navbarIsConsistent() {
        driver.get("http://localhost:3000");
        wait.until(ExpectedConditions.presenceOfElementLocated(By.cssSelector("nav")));
        
        TestivAIWitness.witness(driver, "navbar");
    }
    
    @Test
    void productCardsRender() {
        driver.get("http://localhost:3000");
        
        // Scroll to product cards
        WebElement productSection = wait.until(
            ExpectedConditions.presenceOfElementLocated(By.cssSelector(".product-cards, [data-testid='product-cards']"))
        );
        
        ((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView(true);", productSection);
        
        // Wait for any lazy loading
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        TestivAIWitness.witness(driver, "product-cards");
    }
    
    @Test
    void checkoutFormLayout() {
        driver.get("http://localhost:3000/checkout");
        wait.until(ExpectedConditions.presenceOfElementLocated(By.cssSelector("form")));
        
        TestivAIWitness.witness(driver, "checkout-form");
    }
    
    @Test
    void checkoutFormValidation() {
        driver.get("http://localhost:3000/checkout");
        
        // Fill partial form to trigger validation
        WebElement emailInput = wait.until(
            ExpectedConditions.presenceOfElementLocated(By.cssSelector("#email, [name='email']"))
        );
        emailInput.sendKeys("test@");
        
        // Submit form
        WebElement submitButton = driver.findElement(By.cssSelector("button[type='submit'], [data-testid='submit']"));
        submitButton.click();
        
        // Wait for validation messages
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        TestivAIWitness.witness(driver, "checkout-validation");
    }
}
